Edit pass: grpc batch 3 (#15908)
dotnet/docs: docs/architecture/grpc-for-wcf-developers/self-hosted.md

---
title: Self-hosted gRPC applications - gRPC for WCF developers
description: Deploying ASP.NET Core gRPC applications as self-hosted services.
ms.date: 09/02/2019
---

# Self-hosted gRPC applications

> Although ASP.NET Core 3.0 applications can be hosted in IIS on Windows Server, currently it isn't possible to host a gRPC application in IIS because some of the HTTP/2 functionality isn't supported. This functionality is a goal for a future update to Windows Server.

Author: Saying that the functionality is expected might be construed as a promise to customers, which is risky from a legal standpoint. (If you publicly state that Microsoft will deliver something and then Microsoft doesn’t deliver it, customers can claim that the statement induced them to purchase.) It's safer to tone down or remove statements like that, but I'll restore the original phrasing if you're confident that it won't cause problems.

> [!NOTE]
> If the application isn't running as a Windows service, the `UseWindowsService` method doesn't do anything.

Now publish your application by using one of these methods:

> * From Visual Studio by right-clicking the project and selecting **Publish** on the shortcut menu

Reviewer: ```suggestion
* From Visual Studio by right-clicking the project and selecting **Publish** on the shortcut menu.
```

> * From the .NET Core CLI

Reviewer: ```suggestion
* From the .NET Core CLI.
```

> When you publish a .NET Core application, you can choose to create a *framework-dependent* deployment or a *self-contained* deployment. Framework-dependent deployments require the .NET Core Shared Runtime to be installed on the host where they are run. Self-contained deployments are published with a complete copy of the .NET Core runtime and framework and can be run on any host. For more information, including the advantages and disadvantages of each approach, see the [.NET Core application deployment](https://docs.microsoft.com/dotnet/core/deploying/) documentation.

Reviewer: unrelated to your change to fix link format 

```suggestion
When you publish a .NET Core application, you can choose to create a *framework-dependent* deployment or a *self-contained* deployment. Framework-dependent deployments require the .NET Core Shared Runtime to be installed on the host where they are run. Self-contained deployments are published with a complete copy of the .NET Core runtime and framework and can be run on any host. For more information, including the advantages and disadvantages of each approach, see the [.NET Core application deployment](../../core/deploying/index.md) documentation.
```

dotnet publish -c Release -o ./publish
```

> Copy the complete contents of the `publish` directory to an installation folder. Then, use the [sc tool](https://docs.microsoft.com/windows/desktop/services/controlling-a-service-using-sc) to create a Windows service for the executable file.

Reviewer: ```suggestion
Copy the complete contents of the `publish` directory to an installation folder. Then, use the [sc tool](/windows/desktop/services/controlling-a-service-using-sc) to create a Windows service for the executable file.
```

```console
sc create MyService binPath=C:\MyService\MyService.exe
```

### Log to the Windows event log

> The `UseWindowsService` method automatically adds a [logging](https://docs.microsoft.com/aspnet/core/fundamentals/logging/?view=aspnetcore-3.0) provider that writes log messages to the Windows event log. You can configure logging for this provider by adding an `EventLog` entry to the `Logging` section of `appsettings.json` or another configuration source. 

Reviewer: ```suggestion
The `UseWindowsService` method automatically adds a [logging](/aspnet/core/fundamentals/logging/) provider that writes log messages to the Windows event log. You can configure logging for this provider by adding an `EventLog` entry to the `Logging` section of `appsettings.json` or another configuration source. 
```

You can override the source name used in the event log by setting a `SourceName` property in these settings. If you don't specify a name, the default application name (normally the executable assembly name) will be used.

More information on logging is at the end of this chapter.

> ## Run your app as a Linux service by using systemd

Reviewer 2: I prefer the original phrasing here

Author: I made that change because of this entry in the Microsoft Writing Style Guide: https://styleguides.azurewebsites.net/Styleguide/Read?id=2700&topicid=36373. Is "by using" technically inaccurate?

Reviewer: But that entry also says it's ok to use with product names?

Author: True--though that's the not the case here. I can change the wording back to "with" if you feel strongly about it.

Reviewer: @shirhatti?

Author: I changed it back to "with."

> [!NOTE]
> If the application isn't running as a Linux service, the `UseSystemd` method doesn't do anything.

> Now publish your application. The application can be either framework dependent or self-contained for the relevant Linux runtime (for example, `linux-x64`). You can publish from Visual Studio by right-clicking the project and selecting *Publish* from the shortcut menu. Or you can publish from the .NET Core CLI by using the following command.

Reviewer: I like the change you made earlier that transformed this on a bulleted list. We should do the same here.

Author: Done!

> [!TIP]
> If you have a GUI environment available on your host, a few graphical log viewers are available for Linux, such as *QJournalctl* and *gnome-logs*.

To learn more about querying the `systemd` journal from the command line by using `journalctl`, see [the man pages](https://manpages.debian.org/buster/systemd/journalctl.1).

## HTTPS certificates for self-hosted applications

When you're running a gRPC application in production, you should use a TLS certificate from a trusted certificate authority (CA). This CA might be a public CA, or an internal one for your organization.

> On Windows hosts, you can load the certificate from a secure [certificate store](https://docs.microsoft.com/windows/win32/seccrypto/managing-certificates-with-certificate-stores) by using the [X509Store class](https://docs.microsoft.com/dotnet/api/system.security.cryptography.x509certificates.x509store?view=netcore-3.0). You can also use the `X509Store` class with the OpenSSL key store on some Linux hosts.

Reviewer: ```suggestion
On Windows hosts, you can load the certificate from a secure [certificate store](/windows/win32/seccrypto/managing-certificates-with-certificate-stores) by using the <xref:System.Security.Cryptography.X509Certificates.X509Store> class. You can also use the `X509Store` class with the OpenSSL key store on some Linux hosts.
```

> You can also create certificates by using one of the [X509Certificate2 constructors](https://docs.microsoft.com/dotnet/api/system.security.cryptography.x509certificates.x509certificate.-ctor?view=netcore-3.0), from either:

Reviewer: @shirhatti something looks wrong here. The link text says X509Certificate2 and the link goes to X509Certificate. Which one is correct so I can properly link to the API page using the xref format.

Reviewer 2: It should be `X509Certificate2`

Reviewer: ```suggestion
You can also create certificates by using one of the [X509Certificate2 constructors](xref:System.Security.Cryptography.X509Certificates.X509Certificate2.%23ctor%2A), from either:
```
